cdc/puller/sorter/merger.go

```diff
@@ -417,3 +482,89 @@ func printError(err error) error {
 	}
 	return err
 }
+
+// taskBuffer is used to store pending flushTasks.
+// The design purpose is to reduce the backpressure caused by a congested output chan of the merger,
+// so that heapSorter does not block.
+type taskBuffer struct {
+	mu       sync.Mutex
+	queue    deque.Deque
+	notifier notify.Notifier
+	len      *int64
+	isClosed int32
+}
+
+func newTaskBuffer(len *int64) *taskBuffer {
+	return &taskBuffer{
+		queue:    deque.NewDeque(),
+		notifier: notify.Notifier{},
+		len:      len,
+	}
+}
+
+func (b *taskBuffer) put(task *flushTask) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.queue.PushBack(task)
+	prevCount := atomic.AddInt64(b.len, 1)
+
+	if prevCount == 1 {
+		b.notifier.Notify()
+	}
+}
+
+func (b *taskBuffer) get(ctx context.Context) (*flushTask, error) {
+	if atomic.LoadInt32(&b.isClosed) == 1 && atomic.LoadInt64(b.len) == 0 {
+		return nil, nil
+	}
+
+	if atomic.LoadInt64(b.len) == 0 {
+		recv, err := b.notifier.NewReceiver(time.Millisecond * 50)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		defer recv.Stop()
+
+		startTime := time.Now()
+		for atomic.LoadInt64(b.len) == 0 {
+			select {
+			case <-ctx.Done():
+				return nil, errors.Trace(ctx.Err())
+			case <-recv.C:
+			}
+
+			if atomic.LoadInt32(&b.isClosed) == 1 {
+				return nil, nil
+			}
```

Comment:
The behavior of `get` when `taskBuffer` is closed is not unified.
- In L518, if `b.isClosed == 1` && `b.len == 0`, return nil
- In L518, if `b.isClosed == 1` && `b.len > 0`, will return a task
- In L518, `b.isClosed == 0`, `b.len == 0`, in L537, if `recv.C` bumps and `b.len > 0` and `b.isClosed == 1`, still return nil

Reply:
Besides, can we simplify the check condition, seems the check of `b.len > 0` and `recv.C` do the same work

Reply:
Note that there might be spurious wake-ups because the receiver has a timer.

Reply:
Yep, but receiver can be configured with non-ticker